Add development chain-spec file for minimal/parachain templates for Omni Node compatibility

EleisonC opened the pull request:
# Description

This PR adds development chain specs for the minimal and parachain templates. [#6334](https://github.com/paritytech/polkadot-sdk/issues/6334)


## Integration

This PR adds development chain specs for the minimal and para chain template runtimes, ensuring synchronization with runtime code. It updates zombienet-omni-node.toml, zombinet.toml files to include valid chain spec paths, simplifying configuration for zombienet in the parachain and minimal template.

## Review Notes

1. Overview of Changes:
- Added development chain specs for use in the minimal and parachain template.
- Updated zombienet-omni-node.toml and zombinet.toml files in the minimal and parachain templates to include paths to the new dev chain specs.

2. Integration Guidance:
**NB: Follow the templates' READMEs from the polkadot-SDK master branch. Please build the binaries and runtimes based on the polkadot-SDK master branch.**
- Ensure you have set up your runtimes `parachain-template-runtime` and `minimal-template-runtime`
- Ensure you have installed the nodes required ie `parachain-template-node` and `minimal-template-node`
- Set up [Zombinet](https://paritytech.github.io/zombienet/intro.html)
- For running the parachains, you will need to install the polkadot `cargo install --path polkadot` remember from the polkadot-SDK master branch.
- Inside the template folders minimal or parachain, run the command to start with `Zombienet with Omni Node`, `Zombienet with minimal-template-node` or `Zombienet with parachain-template-node`

*Include your leftover TODOs, if any, here.*
* [ ] Test the syncing of chain specs with runtime's code.

iulianbarbu commented:
> > Thanks @EleisonC ! Left a few comments for now. Looking forward to the README updates & btw, have you clarified how to make a rust test for the chain specs?
> 
> Hey @iulianbarbu that is what I am looking into today. Do you have any pointers you'd like me to know as I move forward?

From the top of my head I thought about adding an integration test. You could build a `ChainSpec` as described here:  https://github.com/paritytech/polkadot-sdk/blob/master/templates/parachain/node/src/chain_spec.rs#L36, and then call `as_json` -> and then convert it to `serde_json::Value`, and then read the newly added file and load its content as a `serde_json::Value` and then compare the two.

EleisonC commented:
Hey @iulianbarbu one last question it's okay if I placed the **test** folder/file under **the nodes** folder/file. if not is there a better place

iulianbarbu commented:
> Hey @iulianbarbu one last question it's okay if I placed the **test** folder/file under **the nodes** folder/file. if not is there a better place

Hmm, thought about placing it under `runtime`  initially. I think it makes more sense to place it under `runtime` so that if anything changes here, running the tests for the crate would fail because the changes will not be inline with `dev_chain_spec.json`. The node uses the logic defined in the runtime, so it would benefit from the runtime test, but not directly. I don't think we need to add a test on the same lines either because the templates' `node`s aren't depending on this `dev_chain_spec.json` file. Only `OmniNode` depends on this file being in sync with the runtime, and that not because it would fail otherwise, but because it would start an outdated version.

EleisonC commented:
> From the top of my head I thought about adding an integration test. You could build a `ChainSpec` as described here: https://github.com/paritytech/polkadot-sdk/blob/master/templates/parachain/node/src/chain_spec.rs#L36, and then call `as_json` -> and then convert it to `serde_json::Value`, and then read the newly added file and load its content as a `serde_json::Value` and then compare the two.

Hey @iulianbarbu regarding comparing the two I am a bit stuck on what to compare. From my research I would think it's other the **code** or value of **top**
Is it the value of **code**? Because the test fails when I compare the code value of the two chain spec files.
`let test_runtime_code = test_json_value
    //     .get("genesis")
    //     .expect("failed to extract genesis")
    //     .get("runtimeGenesis")
    //     .expect("failed to extract runtimeGenesis from genesis")
    //     .get("code")
    //     .expect("failed to extract code from runtimeGenesis");`
    
    But the **code** value of test_runtime and the existing file do not match.  Hence wondering if it would be the **Top** value to or maybe I am missing a step of conversion somewhere

EleisonC commented:
> > From the top of my head I thought about adding an integration test. You could build a `ChainSpec` as described here: https://github.com/paritytech/polkadot-sdk/blob/master/templates/parachain/node/src/chain_spec.rs#L36, and then call `as_json` -> and then convert it to `serde_json::Value`, and then read the newly added file and load its content as a `serde_json::Value` and then compare the two.
> 
> Hey @iulianbarbu regarding comparing the two I am a bit stuck on what to compare. From my research I would think it's other the **code** or value of **top** Is it the value of **code**? Because the test fails when I compare the code value of the two chain spec files. `let test_runtime_code = test_json_value // .get("genesis") // .expect("failed to extract genesis") // .get("runtimeGenesis") // .expect("failed to extract runtimeGenesis from genesis") // .get("code") // .expect("failed to extract code from runtimeGenesis"); ` But the **code** value of test_runtime and the existing file do not match. Hence wondering if it would be the **Top** value to or maybe I am missing a step of conversion somewhere

I figured out my problem was around decoding a Hex String to bytes. I can now properly traverse the values of the Top section and find the code for that spec file

iulianbarbu commented:
> > > From the top of my head I thought about adding an integration test. You could build a `ChainSpec` as described here: https://github.com/paritytech/polkadot-sdk/blob/master/templates/parachain/node/src/chain_spec.rs#L36, and then call `as_json` -> and then convert it to `serde_json::Value`, and then read the newly added file an

d load its content as a `serde_json::Value` and then compare the two.
> > 
> > 
> > Hey @iulianbarbu regarding comparing the two I am a bit stuck on what to compare. From my research I would think it's other the **code** or value of **top** Is it the value of **code**? Because the test fails when I compare the code value of the two chain spec files. `let test_runtime_code = test_json_value // .get("genesis") // .expect("failed to extract genesis") // .get("runtimeGenesis") // .expect("failed to extract runtimeGenesis from genesis") // .get("code") // .expect("failed to extract code from runtimeGenesis"); ` But the **code** value of test_runtime and the existing file do not match. Hence wondering if it would be the **Top** value to or maybe I am missing a step of conversion somewhere
> 
> I figured out my problem was around decoding a Hex String to bytes. I can now properly traverse the values of the Top section and find the code for that spec file

Happy to hear you've sorted it out! I've noticed that in some tests we don't compare the code portions of the generated chain specs, and that's mainly because they are not super relevant for the test case. For this test though I think that it is relevant, to ensure generated chain specs correspond to the runtimes in the CI.

Some other good news is that once we'll have these chain specs we'll be able to easily run in the CI the `omni_node_*` tests like this one: https://github.com/paritytech/polkadot-sdk/blob/master/templates/zombienet/tests/smoke.rs#L160. We can update them in a follow up, no need to consider them right now, but it is up to you.

As usual, LMK in case of blockers, happy to take a look.

EleisonC commented:
Hey @iulianbarbu, 
I added a test and would appreciate it if you could look at it and approve it before I add the one for the para chain. Additionally, I added the test folder to nodes because I was encountering setup issues.

iulianbarbu reviewed (commented):
Can you please regenerate the `dev_chain_spec.json` for minimal (for now) after building the `minimal/runtime` with `cargo build`? I wonder if you built the runtime with `--release` and then generated the `dev_chain_spec.json`. It makes sense according to the README, but we run the tests with the `testnet` profile, which enables the debug symbols, and I think the chain spec `code` section differs based on the profile. I would be curious to run the test again after you regenerate the `dev_chain_spec.json`, and also see if regenerating on my machine results in a different code section, or chain spec altogheter.

EleisonC commented on templates/minimal/node/tests/validate_dev_chain_spec.rs:
```diff
@@ -0,0 +1,84 @@
+use codec::Decode;
```
Done

EleisonC commented:
> Can you please regenerate the `dev_chain_spec.json` for minimal (for now) after building the `minimal/runtime` with `cargo build`? I wonder if you built the runtime with `--release` and then generated the `dev_chain_spec.json`. 

Hey @iulianbarbu  Yes, I ran the cargo build with `--release`, and I also added the newly generated Spec file. Lastly, I relocated the test directory to the runtime folder.

iulianbarbu commented on templates/minimal/runtime/tests/validate_dev_chain_spec.rs:
```diff
@@ -0,0 +1,87 @@
+	let existing_chain_spec: serde_json::Value =
+		serde_json::from_reader(existing_chain_spec_reader).expect("should read proper JSON. qed");
+
+    let data1 = decode_runtime_info(existing_chain_spec).unwrap();
+    let data2 = decode_runtime_info(test_chain_spec).unwrap(); 
```
My intention was to avoid these, and assert directly on the JSONs equivalence. Does the test fail on your end if not adding these, when comparing the JSONs? It might be because of the difference that I mentioned. The test needs the chain_specs to be generated based on a runtime built with `cargo build`, without the `--release`profile.

EleisonC replied:
My bad  🤦🏾. I am going to go and do it with out the --release

EleisonC replied:
Hey @iulianbarbu  please correct me where I am wrong.
My understanding of the chain specification, particularly the fields, started from this link: 

https://paritytech.github.io/polkadot-sdk/master/sc_chain_spec/index.html#genesis-initial-runtime-state. 

Specifically, I looked at the genesis, which includes the Code or the top JSON representation of the chain specification found at https://paritytech.github.io/polkadot-sdk/master/sc_chain_spec/index.html. 

Since the test chain specification is created with the name `with_name("test_Development")`, simply comparing the JSONs may not yield clear results. Therefore, I traversed the JSON structure to locate the code value, which I can then use for comparison.

Therefore, I use `sp_version::RuntimeVersion` and pass it the code value to access runtime API, specifically the `spec_name`, `spec_version`, and `authoring_version` as referenced in https://paritytech.github.io/polkadot-sdk/master/sp_version/struct.RuntimeVersion.html.

EleisonC replied:
> Does the test fail on your end if not adding these, when comparing the JSONs?

Yes it does fail

iulianbarbu replied:
> Since the test chain specification is created with the name with_name("test_Development"), simply comparing the JSONs may not yield clear results.

You are right. Sorry for the confusion, I missed changing this when suggested the change. The idea is to modify `with_name`, `with_id` & `chain_type` to be the same as those inside the `dev_chain_spec.json`. You can open the `dev_chain_spec.json` and see the values for `name`, `id`, and `chain_type`. And then, combined with the idea of generating the `dev_chain_spec.json` with `cargo build` instead of `cargo build --release`, the created chain spec in the test converted to a  `serde_json::Value` should be the same with the `serde_json::Value` read from `dev_chain_spec.json`.

EleisonC replied:
I see that test approaching  🟢
Making a push with those specifications

EleisonC replied:
Hey @iulianbarbu 
I added the test, and it passed successfully. Everything is working as expected.
Let me proceed to create the one for the parachain I guess unless there are other changes to look at

iulianbarbu reviewed (commented):
Thanks @EleisonC ! I think it is almost there. Just want to ensure you're building the runtimes without `--release`, just with `cargo build`, and then generate the chain specs based on them.

iulianbarbu commented on templates/minimal/runtime/tests/validate_dev_chain_spec.rs:
```diff
@@ -0,0 +1,59 @@
+	let existing_chain_spec: serde_json::Value =
+		serde_json::from_reader(existing_chain_spec_reader).expect("should read proper JSON. qed");
+
+    assert_eq!(existing_chain_spec, test_chain_spec);
```
As I was afraid, running the test on my machine failed. Confirmed with a few more people that building the runtimes isn't deterministic yet. We could use `srtool`: https://github.com/chevdor/srtool to get deterministic runtimes and as a result deterministic `dev_chain_spec.json`s. There are two approaches from what I can tell:

1. Use `srtool-cli` to generate the `dev_chain_spec.json` locally, and when it comes to the test that should run in the CI, we'd rely on a github action that installs srtool-cli, runs it against the runtimes to build the wasm blob, and then installs `chain-spec-builder` and uses it against the built `.wasm` to generate the `dev_chain_spec.json`. Then finally runs a diff between the CI `dev_chain_spec.json` and the one from the templates.

2. Use `srtool-action` as we use it here: https://github.com/paritytech/polkadot-sdk/blob/master/.github/workflows/release-30_publish_release_draft.yml#L53. IMO it is not doing exactly what we want since it favors cases where we build the majority of the existing runtimes (because of the `excluded runtimes` logic, which is easier to use when the excluded runtimes list is small, or known), instead of building only certain known runtimes. IMO it needs some adjustments to make it work easily for our use case too, and then once the runtimes are built (and corresponding wasm blobs are uploaded as artifacts), download the artifacts, install `chain-spec-builder`, generate the `dev_chain_spec.json` and compare it with the existing `dev_chain_spec.json`, as in the first case.

Neither of the options is straightforward, and as you can notice, we'd change the direction from a Rust test to a github workflow. Doing either of the options could be slow too given the GH actions that run on this PR need to be approved by a maintainer first, and that impacts your feedback cycle loop. I know it might be disappointing coming with such a big request for a change at this late hour, and I get it if you'd like to stop here for now, but if you can still invest time in this LMK, I would be happy to assist.

EleisonC replied:
Hey @iulianbarbu, I will implement the strategies you suggested. I will reach out if I have any questions.

EleisonC replied:
Writing the integration test, was a dope opportunity to learn the product⚡️.  How the different components interact.  I enjoyed it. Great onboarding task

EleisonC commented:
Hey @iulianbarbu  I have uploaded a new dev_spec_chain.json using srtool runtime. The next step should be creating a GitHub workflow.

Are there any templates that are used to or usually build the workflows or they are built from scratch?

I found a useful resource. I could borrow something, but for our purpose, we will be working with the `srtool-cli`.  
https://github.com/chevdor/srtool-actions

iulianbarbu commented:
> Are there any templates that are used to or usually build the workflows or they are built from scratch?

Usually built from scratch, but you can find much inspiration in existing workflows from .github directory.

> I found a useful resource. I could borrow something, but for our purpose, we will be working with the srtool-cli.
> https://github.com/chevdor/srtool-actions

srtool-actions is a good choice. I haven't used any of the two options yet but as I told you in a previous comment, we're already using srtool-actions in our CI. Feel free to pick any of the two options though, I don't have a strong preference.

EleisonC commented:
Hey @iulianbarbu, I added the GitHub workflow to validate the template minimal dev_chain_spec.json. Please let me know if you have any feedback before I add one for the parachains.

iulianbarbu commented on .github/workflows/check-runtimes-template-spec-chain.yml:
```diff
@@ -0,0 +1,135 @@
+name: check-runtimes-template-spec-chain.yml
```
I think this should be named to suggest it is related to minimal template, as you did for parachain template.

iulianbarbu replied:
I think we can create a custom workflow that receives parameters (mainly the package name and the associated runtime path), so that we don't have duplicate workflows for parachain/minimal templates. WDYT?

EleisonC replied:
I definitely agree with you on that. I first wanted to see the workflow pass 😅. To much cc cv

iulianbarbu commented on .github/workflows/check-runtimes-template-spec-chain.yml:
```diff
@@ -0,0 +1,135 @@
+  workflow_dispatch:
+
+jobs:
+  preflight:
+    uses: ./.github/workflows/reusable-preflight.yml
```
I don't think we need this. Seeing the preflight used specifically when needing to run this on custom runners for certain cases (like release activities). We should be fine with `ubuntu-latest` for the jobs bellow.

iulianbarbu commented on .github/workflows/check-runtimes-template-spec-chain.yml:
```diff
@@ -0,0 +1,135 @@
+
+      - name: Pull srtool docker image
+        run: |
+          srtool pull
```
Looks like it can not detect the container engine. I am seeing an error here: https://github.com/paritytech/polkadot-sdk/actions/runs/12244392003/job/34180482077?pr=6529. It is probably because the runner where this runs is not configured with `docker` or `podman`.

iulianbarbu replied:
The job still fails: https://github.com/paritytech/polkadot-sdk/actions/runs/12254440065/job/34185795791?pr=6529. We need to use the `ubuntu-latest` runner for this and drop the reusable preflight runner.

iulianbarbu reviewed (commented):
This is nice, it goes into the right direction I think!

EleisonC commented:
Had not seen the comments on them right now

iulianbarbu commented on .github/workflows/verify-minimal-template-spec-chain.yml:
```diff
@@ -0,0 +1,139 @@
+
+      - name: Build your runtime binary
+        run: |
+          srtool build --package minimal-template-runtime --runtime-dir ./templates/minimal/runtime
```
I think this picks `podman` by defualt, which runs rootless, and you'll need to do something like:
```suggestion
          srtool build --package minimal-template-runtime --runtime-dir ./templates/minimal/runtime --root
```

Same would be for the parachain job too.

EleisonC commented:
Made the changes to the new combined Workflow @iulianbarbu

alvicsam commented on .github/workflows/veriy-minimal-parachain-dev-spec-chains.yml:
```diff
@@ -0,0 +1,137 @@
+      - name: Checkout code
+        uses: actions/checkout@v4
+
+      - name: Install Rust
+        uses: actions-rs/toolchain@v1
+        with:
+          profile: minimal
+          toolchain: stable
```
Please install rust as it's done [here](https://github.com/paritytech/polkadot-sdk/blob/2dd2bb5a847d7bf006493d9fe9b10a13289118b6/.github/workflows/checks-quick.yml#L191-L202) (taking toolchain version from `.github/env`)

EleisonC commented on .github/workflows/veriy-minimal-parachain-dev-spec-chains.yml:
```diff
@@ -0,0 +1,137 @@
+
+permissions:
+  contents: read
+  actions: write
```
My thought was that some steps might require writing (or creating) 
`      - name: Prepare build directories
        run: |
          sudo mkdir -p ${{ matrix.template.runtime_path }}/target
          sudo chmod -R 777 ${{ matrix.template.runtime_path }}/target
`

creating new files that was my reasoning. What are your thoughts on this? Was I on the right path?
@alvicsam

alvicsam replied:
Not exactly, I'd remove it since `actions: write` doesn't affect creating files/folders inside job in a workflow.

EleisonC commented:
@iulianbarbu  Update: I am trying to run the workflow steps locally using Podman and Srtool, as I hope to compare the wasm binary (spec chain.json file) that I build locally. Do you think the Rust version used in the workflow could have any impact? I'm just curious. I realize that’s what Srtool is designed to address, along with the differences between Podman and Docker that I'm attempting to replicate. However, I want to be confident about this process.

iulianbarbu commented:
Hi @EleisonC ! I am expecting the chain specs to be the same if generated with any 'chain-spec-builder' version (no matter how it got built).

Can you compare the wasm blob generated on the CI with the one generated on your machine? Alternatively, if they differ, I would compare wasm runtimes generated by two runs of the GH action you introduced (by downloading manually the wasms from action storage, if you have access to the action output). I am assuming they will be the same at least when built on the CI for multiple times from the same git commit.

@skunert any idea why the chain specs differ even if both were built based on the same runtime, built with srtool?

skunert commented:
This is not related to the rust version or the chainspec itself. The runtime wasm is the only changing factor here. The rust version is handled by the docker image pulled by SRTool.

What happened here is different. When building this branch locally I arrived at this summary from SRTool (parachain template runtime):
```
✨ Your Substrate WASM Runtime is ready! ✨
✨ WASM  : ./templates/parachain/runtime/target/srtool/release/wbuild/parachain-template-runtime/parachain_template_runtime.compact.wasm
✨ Z_WASM: ./templates/parachain/runtime/target/srtool/release/wbuild/parachain-template-runtime/parachain_template_runtime.compact.compressed.wasm
Summary generated with srtool v0.17.0 using the docker image paritytech/srtool:1.81.0:
 Package     : parachain-template-runtime v0.0.0
 GIT commit  : 546ba5655f60f9831e99fda3d2b577ddf40205ef
 GIT tag     : v1.16.0-approval-voting
 GIT branch  : add-dev-chain-spec
 Rustc       : rustc 1.81.0 (eeb90cda1 2024-09-04)
 Time        : 2024-12-17T14:36:18Z

== Compact
 Version          : parachain-template-runtime-1 (parachain-template-runtime-0.tx1.au1)
 Metadata         : V14
 Size             : 4.41 MB (4621253 bytes)
 setCode          : 0x1548c48d31461c75f3ea76e6ae61f54f575ce077ab41bee260782a073cdb5f3f
 authorizeUpgrade : 0x55d46074a49ca56692ec31375746f0f746a5b0f649d41dbcdbeb70986dce9b0b
 IPFS             : QmaJMVKPqAcG9W4H3Q4KJBnXM7kzZFJBzF3wNcoLEBJZu9
 BLAKE2_256       : 0x3d18a9732ea4e61285a9e5e33632109af7e2e84e7184a5963009c0c6fd649e41
 Wasm             : ./templates/parachain/runtime/target/srtool/release/wbuild/parachain-template-runtime/parachain_template_runtime.compact.wasm

== Compressed
 Version          : parachain-template-runtime-1 (parachain-template-runtime-0.tx1.au1)
 Metadata         : V14
 Size             : 963.69 kB (986814 bytes)
 Compression      : 78.65%
 setCode          : 0xa01a1db41163f56cff8838ce27001157a08919061e159bcdf57f1df7b55e5416
 authorizeUpgrade : 0xde0908e96c72608b3b392b34997ec707de13c91df924fe9815faede67b35a6cf
 IPFS             : QmbdeeheW2ZjsxmNr8dAhbq67CLsGN1E4G7boUNZ4b5uJs
 BLAKE2_256       : 0xc5e7b355e18c5a77fc8ced1ef228631b504e81cde98aea1b332a7df086809926
 Wasm             : ./templates/parachain/runtime/target/srtool/release/wbuild/parachain-template-runtime/parachain_template_runtime.compact.compressed.wasm
```

We see that the Blake2 hash of the runtime is `0xc5e7b355e18c5a77fc8ced1ef228631b504e81cde98aea1b332a7df086809926`. In contrast in this line of the [CI Job](https://github.com/paritytech/polkadot-sdk/actions/runs/12327187654/job/34410440475?pr=6529#step:8:1050)  we see that the hash of the runtime build by CI is `0xd7f7a9dd1ab203a0df5237c643844cfdaa64521320f835b5f32404fa99c7be40`.

The reason is that CI merges the current branch with master for the job. Master had some recent changes to the parachain template runtime which resulted in a different WASM blob. After merging master I arrive at the same hash as CI:
```
✨ Your Substrate WASM Runtime is ready! ✨
✨ WASM  : ./templates/parachain/runtime/target/srtool/release/wbuild/parachain-template-runtime/parachain_template_runtime.compact.wasm
✨ Z_WASM: ./templates/parachain/runtime/target/srtool/release/wbuild/parachain-template-runtime/parachain_template_runtime.compact.compressed.wasm
Summary generated with srtool v0.17.0 using the docker image paritytech/srtool:1.81.0:
 Package     : parachain-template-runtime v0.0.0
 GIT commit  : 546ba5655f60f9831e99fda3d2b577ddf40205ef
 GIT tag     : v1.16.0-approval-voting
 GIT branch  : add-dev-chain-spec
 Rustc       : rustc 1.81.0 (eeb90cda1 2024-09-04)
 Time        : 2024-12-17T18:54:59Z

== Compact
 Version          : parachain-template-runtime-1 (parachain-template-runtime-0.tx1.au1)
 Metadata         : V14
 Size             : 4.41 MB (4621228 bytes)
 setCode          : 0xbfe2d975e536b433bdd4883da768d6bf538eda1670f8bd7373c0dfdbabb76f14
 authorizeUpgrade : 0xa96a217a964c9734620e430081adfffa4abf971b66e5fb60969b18878c464bee
 IPFS             : QmYSNTQisEr3gyxzxFDhmg1XugwknQ4njjRZ4bUKvNbaLQ
 BLAKE2_256       : 0xf9ce00caa1b5d5a9b5d5d464bd13caf5b1b821248a8695ec28edf99f5adb2449
 Wasm             : ./templates/parachain/runtime/target/srtool/release/wbuild/parachain-template-runtime/parachain_template_runtime.compact.wasm

== Compressed
 Version          : parachain-template-runtime-1 (parachain-template-runtime-0.tx1.au1)
 Metadata         : V14
 Size             : 963.72 kB (986850 bytes)
 Compression      : 78.65%
 setCode          : 0x9bb53a5bb708fceb70940a1fd0de435eb2ecc08b0748cd3840fad3e69b68544e
 authorizeUpgrade : 0xf0791b1c05bf46e6ce32bf452ce20966b62f9a4937911948312265b8914ec14c
 IPFS             : QmaUPBrdJgx7VxeVgHDTnAL8541eTfxHCFgyMSLNDEWECq
 BLAKE2_256       : 0xd7f7a9dd1ab203a0df5237c643844cfdaa64521320f835b5f32404fa99c7be40
 Wasm             : ./templates/parachain/runtime/target/srtool/release/wbuild/parachain-template-runtime/parachain_template_runtime.compact.compressed.wasm
```

EleisonC commented:
Hey @iulianbarbu I uploaded a new dev_chain_spec for Parachain. I checked it against the out come of the previous run, and it looks hopeful. Getting around the minimal one

EleisonC commented:
> We see that the Blake2 hash of the runtime

@iulianbarbu  also I still get different hashes for my minimal. Clearing my cache and all to give it another try

EleisonC commented:
Hey @skunert tried the same with the minimal but I keep on getting a BLAKE2_256 ` BLAKE2_256       : 0x5a8dacd160f12a36fd0a7257606283baefbba77104cd67ced0b567b4407aa817 ` value  that is not exact as the CI run 
```
✨ Your Substrate WASM Runtime is ready! ✨
✨ WASM  : ./templates/minimal/runtime/target/srtool/release/wbuild/minimal-template-runtime/minimal_template_runtime.compact.wasm
✨ Z_WASM: ./templates/minimal/runtime/target/srtool/release/wbuild/minimal-template-runtime/minimal_template_runtime.compact.compressed.wasm
fatal: No names found, cannot describe anything.
Summary generated with srtool v0.17.0 using the docker image paritytech/srtool:1.81.0:
 Package     : minimal-template-runtime v0.0.0
 GIT commit  : 278324c7ff3acc968adcf414b52507111b3e68d4
 GIT tag     : 
 GIT branch  : add-dev-chain-spec
 Rustc       : rustc 1.81.0 (eeb90cda1 2024-09-04)
 Time        : 2024-12-19T17:07:49Z

== Compact
 Version          : minimal-template-runtime-0 (minimal-template-runtime-1.tx1.au1)
 Metadata         : V14
 Size             : 1004.33 kB (1028437 bytes)
 setCode          : 0x8dc8fbbdbc4fb058e076d915175b9363091f9af13cbaf2e78cdaa9bd6f94dc95
 authorizeUpgrade : 0xe91aed2e617f6243522dab081573bb50292877970bc83146e004eb8613cb0cc4
 IPFS             : QmU5nuekV7ca4ShAFU2uTgzz13jzEW3nAfPcwEFPFGHWsV
 BLAKE2_256       : 0xf80020328d1afd7bbb88effab38bbfdaab67868cc7e94c66abe4e10da17e96e5
 Wasm             : ./templates/minimal/runtime/target/srtool/release/wbuild/minimal-template-runtime/minimal_template_runtime.compact.wasm

== Compressed
 Version          : minimal-template-runtime-0 (minimal-template-runtime-1.tx1.au1)
 Metadata         : V14
 Size             : 262.30 kB (268593 bytes)
 Compression      : 73.89%
 setCode          : 0x6a57cc079da9f700700e64e55eb30d765b06caee2b307c10229d67eaf36cfac8
 authorizeUpgrade : 0x704531ce06cacf0c133e9f8644c534930509045848c9d8507dc626eb2c98e9d7
 IPFS             : QmXWdrLUvaM7ZQ9F6bUzKxUADfhWojU1y1MNichrozFci2
 BLAKE2_256       : 0x5a8dacd160f12a36fd0a7257606283baefbba77104cd67ced0b567b4407aa817
 Wasm             : ./templates/minimal/runtime/target/srtool/release/wbuild/minimal-template-runtime/minimal_template_runtime.compact.compressed.wasm
``` I cleared my cache and ran the command with the latest master, but I'm still building that WASM blob. Any pointers?

skunert commented:
I am getting the same as your posted snippet from CI:
```
✨ Your Substrate WASM Runtime is ready! ✨
✨ WASM  : ./templates/minimal/runtime/target/srtool/release/wbuild/minimal-template-runtime/minimal_template_runtime.compact.wasm
✨ Z_WASM: ./templates/minimal/runtime/target/srtool/release/wbuild/minimal-template-runtime/minimal_template_runtime.compact.compressed.wasm
Summary generated with srtool v0.17.0 using the docker image paritytech/srtool:1.81.0:
 Package     : minimal-template-runtime v0.0.0
 GIT commit  : f0a3f18168f27f66f1cab052b402b67b0f4a5dd7
 GIT tag     : v1.16.0-approval-voting
 GIT branch  : add-dev-chain-spec
 Rustc       : rustc 1.81.0 (eeb90cda1 2024-09-04)
 Time        : 2024-12-19T18:25:29Z

== Compact
 Version          : minimal-template-runtime-0 (minimal-template-runtime-1.tx1.au1)
 Metadata         : V14
 Size             : 1004.33 kB (1028437 bytes)
 setCode          : 0x8dc8fbbdbc4fb058e076d915175b9363091f9af13cbaf2e78cdaa9bd6f94dc95
 authorizeUpgrade : 0xe91aed2e617f6243522dab081573bb50292877970bc83146e004eb8613cb0cc4
 IPFS             : QmU5nuekV7ca4ShAFU2uTgzz13jzEW3nAfPcwEFPFGHWsV
 BLAKE2_256       : 0xf80020328d1afd7bbb88effab38bbfdaab67868cc7e94c66abe4e10da17e96e5
 Wasm             : ./templates/minimal/runtime/target/srtool/release/wbuild/minimal-template-runtime/minimal_template_runtime.compact.wasm

== Compressed
 Version          : minimal-template-runtime-0 (minimal-template-runtime-1.tx1.au1)
 Metadata         : V14
 Size             : 262.30 kB (268593 bytes)
 Compression      : 73.89%
 setCode          : 0x6a57cc079da9f700700e64e55eb30d765b06caee2b307c10229d67eaf36cfac8
 authorizeUpgrade : 0x704531ce06cacf0c133e9f8644c534930509045848c9d8507dc626eb2c98e9d7
 IPFS             : QmXWdrLUvaM7ZQ9F6bUzKxUADfhWojU1y1MNichrozFci2
 BLAKE2_256       : 0x5a8dacd160f12a36fd0a7257606283baefbba77104cd67ced0b567b4407aa817
 Wasm             : ./templates/minimal/runtime/target/srtool/release/wbuild/minimal-template-runtime/minimal_template_runtime.compact.compressed.wasm
```

Did you maybe mix up your build command? I am using `srtool build --package minimal-template-runtime --runtime-dir ./templates/minimal/runtime --root` on your branch here. Please stop merging master, I think your are having a different issue this time.

EleisonC commented:
> Did you maybe mix up your build command? I am using `srtool build --package minimal-template-runtime --runtime-dir ./templates/minimal/runtime --root` on your branch here.

sorry should have added the commands `srtool build --package minimal-template-runtime --runtime-dir ./templates/minimal/runtime --root` is what I used same as the one you used @skunert

skunert commented:
Hmm at this point its not really possible for me to say why you are ending up with a different hash. Have you merged master here in the PR but not pulled to your local machine, so that you are on a different commit? Or your have local changes?

These issues aside, after thinking about it more I am against introducing this CI job. Everytime there is a minor change in any of the runtime related code of polkadot-sdk, the generated WASM blob will change, forcing developers to update this chain-spec constantly (because the job will fail). 

A better way to handle this IMO is to just have the chainspec committed and update it manually. For sure this is annoying and the spec could potentially go out of sync. But having an extra CI job for it seems like overkill.

I am sorry for the work you have already put into the CI job. Can you please remove it?

EleisonC commented:
> Hmm at this point its not really possible for me to say why you are ending up with a different hash. Have you merged master here in the PR but not pulled to your local machine, so that you are on a different commit? Or your have local changes?

hey I appreciate the help and the time you were able to allocate to this. I want to first address this issue ☝️, After thorough testing, I can confirm that this issue has been resolved and the workflows run okay [Workflow Run Personal Fork](https://github.com/EleisonC/polkadot-sdk/actions/runs/12423286460). The CI job and local development environments now consistently produce the same Blake2_256 hash for the runtime WASM blob, ensuring alignment with the Polkadot SDK master

EleisonC commented:
> Hmm at this point its not really possible for me to say why you are ending up with a different hash. Have you merged master here in the PR but not pulled to your local machine, so that you are on a different commit? Or your have local changes?

Hey @skunert cc: @iulianbarbu .  I appreciate the help and the time you were able to allocate to this. I want to first address this issue ☝️, After thorough testing, I can confirm that this issue has been resolved and the workflows run okay [Workflow Run Personal Fork](https://github.com/EleisonC/polkadot-sdk/actions/runs/12423286460). The CI job and local development environments now consistently produce the same Blake2_256 hash for the runtime WASM blob, ensuring alignment with the Polkadot SDK master's Runtime determined using the Srtool. 
cmd-ran srtool build --package minimal-template-runtime --runtime-dir ./templates/minimal/runtime --root


> These issues aside, after thinking about it more I am against introducing this CI job. Everytime there is a minor change in any of the runtime related code of polkadot-sdk, the generated WASM blob will change, forcing developers to update this chain-spec constantly (because the job will fail).
> 
> A better way to handle this IMO is to just have the chainspec committed and update it manually. For sure this is annoying and the spec could potentially go out of sync. But having an extra CI job for it seems like overkill.
> 
> I am sorry for the work you have already put into the CI job. Can you please remove it?

I completely understand your concern regarding the added maintenance burden of a CI job that enforces frequent updates to the chain spec due to minor runtime changes. Your suggestion to commit the chain spec manually and update it as needed is reasonable; however, I'd like to propose an enhancement that maintains the benefits of automation while keeping the workflow lightweight and developer-friendly.

1. The CI pipeline can include a lightweight job
Add a job to verify if the code field inside the genesis section of the committed chain spec matches the corresponding field in the CI-generated spec file. If there's a mismatch, the job should fail with clear, actionable guidance for developers to update the chain spec. This balances automation and developer effort without introducing unnecessary overhead. This could be use full in situation where a runtime was probably build the wrong way or with outdated version of staging or diff srtool version used and might not realise.

2. Introduce a Shell script for easier updates
Provide a shell script to streamline the process of installing srtool, pulling updates, building, and updating the dev chain spec code. This simplifies the workflow for developers, requiring only a single executable shell script to update only requiring them to then push changes. 
By leveraging this, the CI workflow can serve as an attestation mechanism to ensure that the dev chain is built using the same runtime as Polkadot SDK.

skunert commented:
Hey,

the problem with (1) is that this will not break now and then when there is a change to the parachain [template runtime](https://github.com/paritytech/polkadot-sdk/blob/f0a3f18168f27f66f1cab052b402b67b0f4a5dd7/templates/parachain/runtime/src/lib.rs), that would be okay. But polkadot-sdk is merging lots of commits per week. Imagine someone is creating a pull request to modify some log line [here](https://github.com/paritytech/polkadot-sdk/blob/f0a3f18168f27f66f1cab052b402b67b0f4a5dd7/substrate/frame/system/src/extensions/check_weight.rs#L59-L59) for example. In their pull request the CI job would fail because that code is a dependency of parachain-template-runtime and they would need to regenerate the chain-spec.

Whe question is also what the benefit of keeping the chain-spec up to date in polkadot-sdk really is. Once per release these two template repositories are synchronized with the current state of polkadot-sdk:
- polkadot-parachain-template: https://github.com/paritytech/polkadot-sdk-parachain-template
- polkadot-minimal-template: https://github.com/paritytech/polkadot-sdk-minimal-template

This is where I expect new users to go to try out these templates. Ideally, the [job](https://github.com/paritytech/polkadot-sdk/blob/4e805ca05067f6ed970f33f9be51483185b0cc0b/.github/workflows/misc-sync-templates.yml#L6-L6) that synchronizes the templates with the polkadot-sdk would generate the chain-specs from the current runtime. This way, the chain-specs in the template repositories would be correct on every release.

I like your proposal of (2), having a script in the repo to generate these chainspecs. The flow could look like this:
- Have a manually generated chain-spec in polkadot-sdk
- Have a script that refreshes the chain-spec in polkadot-sdk (gets copied over to the template repos)
- The job that syncs the code from here into the template repositories regenerates the chain-spec and makes sure that it is up-to-date in the template repos.

This way the chain-spec in the template repositories is always up-to-date and we don't have the burden of this constantly failing CI job.

iulianbarbu commented:
Hey @EleisonC ! First of all, sorry for my silence, it has been the end of the year and I took some needed time off. 

Second of all, I have to admit I totally missed the aspect of this CI job failing at every minor change done in `polkadot-sdk`, in any crate which is a dependency of any of the two runtimes (`minimal` or `parachain`). At this point I am leaning towards @skunert idea of dropping the CI job, considering it will bring attention with most PRs and its benefit - ensuring the chain specs that come with the templates are up to do for usage of templates based on `polkadot-sdk` repo, when we aim in fact for usage from their dedicated GH repos (e.g. https://github.com/paritytech/polkadot-sdk-parachain-template). 

Generating the chain specs once in this job: https://github.com/paritytech/polkadot-sdk/blob/master/.github/workflows/misc-sync-templates.yml should be enough IMO, from the point of view of ensuring up to date chain specs (in relation to the code found in their repo), and ease the use when it comes to `OmniNode`. LMK if you think differently.
I also believe the CI job effort can conclude nicely in this script you're proposing, which can be used anytime we want to update the chain specs for `polkadot-sdk` (as a nice to have, we can also print some notes right before the script is doing its thing, about the merging of master with the branch for the case of a PR, in case we want to update the chain spec as it would be done in CI), or when we generate the chain specs in the misc-sync-templates action. This script can install `srtool` if not present and check the environment looks fine (having at least `podman` installed). The script can be added as a follow up though (but it is your choice, given the `misc-sync-templates.yml` would contain similar commands, and could very well use a script that can do the same separately from the action).

Lastly, wanted to say that your work here was a great learning experience for me too. Would be happy to assist and review any followup changes if you can invest some more time here. LMK your thoughts.

EleisonC commented:
> Hey @EleisonC ! First of all, sorry for my silence, it has been the end of the year and I took some needed time off.

Hey @iulianbarbu I agree the time-off it is needed. I did the same. It was Christmas I bet it was enough haha. Information overload can happen quickly.  😅

> Second of all, I have to admit I totally missed the aspect of this CI job failing at every minor change done in polkadot-sdk, in any crate which is a dependency of any of the two runtimes (minimal or parachain). At this point I am leaning towards @skunert idea of dropping the CI job, considering it will bring attention with most PRs and its benefit - ensuring the chain specs that come with the templates are up to date for usage of templates based on polkadot-sdk repo, when we aim in fact for usage from their dedicated GH repos (e.g. https://github.com/paritytech/polkadot-sdk-parachain-template).
> 
> Generating the chain specs once in this job: https://github.com/paritytech/polkadot-sdk/blob/master/.github/workflows/misc-sync-templates.yml should be enough IMO, from the point of view of ensuring up to date chain specs (in relation to the code found in their repo), and ease the use when it comes to OmniNode. LMK if you think differently.

Based on this my Next steps are going to be to : 
- Remove the current CI job that verifies chain specs
- Add chain spec generation to the misc-sync-templates.yml workflow 
    - Generate specs only during template synchronization

> I also believe the CI job effort can conclude nicely in this script you're proposing, which can be used anytime we want to update the chain specs for polkadot-sdk (as a nice to have, we can also print some notes right before the script is doing its thing, about the merging of master with the branch for the case of a PR, in case we want to update the chain spec as it would be done in CI), or when we generate the chain specs in the misc-sync-templates action. This script can install srtool if not present and check the environment looks fine (having at least podman installed). The script can be added as a follow up though (but it is your choice, given the misc-sync-templates.yml would contain similar commands, and could very well use a script that can do the same separately from the action).

This can be done in another PR. Its still hints at one of the earlier solutions 
Create a utility script that:
Installs srtool if not present
Verifies at least podman installation
Generates chain specs on demand
Prints informative notes about branch merging
Can be used both manually and by the sync workflow


> Lastly, wanted to say that your work here was a great learning experience for me too. Would be happy to assist and review any followup changes if you can invest some more time here. LMK your thoughts.

This is work I am proud of and definitely committed to seeing through. Also, I appreciate your continued guidance and support it makes the process easier.

EleisonC commented:
Hey @iulianbarbu, I’d love to hear your thoughts on my approach and whether it aligns with the expectations.
cc: @skunert – I would appreciate your perspective as well.

iulianbarbu commented on .github/workflows/misc-sync-templates.yml:
```diff
@@ -27,7 +27,63 @@ on:
+            srtool build --package ${{ matrix.package_name }} --runtime-dir ${{ matrix.runtime_path }} --root
+            
+            # Generate chain spec
+            chain-spec-builder -c dev_chain_spec.json create \
+              --relay-chain "${{ matrix.relay_chain }}" \
+              --para-id 1000 \
+              --runtime "${{ matrix.runtime_path }}/target/srtool/release/wbuild/${{ matrix.runtime_wasm_path }}" \
+              named-preset development
```
nit:
```suggestion
            # Note that para-id is set to 1000 for both minimal/parachain templates. 
            # `parachain-runtime` is hardcoded to use this parachain id. 
            # `minimal` template isn't using it, but when started with Omni Node, this para id is required (any number can do it, so setting it to 1000 for convenience).
            chain-spec-builder -c dev_chain_spec.json create \
              --relay-chain "${{ matrix.relay_chain }}" \
              --para-id 1000 \
              --runtime "${{ matrix.runtime_path }}/target/srtool/release/wbuild/${{ matrix.runtime_wasm_path }}" \
              named-preset development
```

iulianbarbu commented on templates/minimal/README.md:
```diff
@@ -37,11 +37,25 @@
 template for building a blockchain node.
 
+
+- 🔧 To simplify the process of using Zombienet with Omni Node, we've added a pre-configured development
+chain spec (dev_chain_spec.json) to the minimal template. Additionally, the zombienet-omni-node.toml file
+has been updated to include a valid chain spec path pointing to dev_chain_spec.json
```
I would move this to the `#### Update `zombienet-omni-node.toml` with a valid chain spec path` section, right before the network spec sample where we instruct where the chain_spec_path updating should be done.

```suggestion
- 🔧 To simplify the process of starting the minimal-template with Zombienet and Omni Node, we've added a pre-configured development
chain spec (dev_chain_spec.json) to the minimal template. The `zombienet-omni-node.toml` file of this template points to it, but you can update it to an updated chain spec generated on your machine, as shown below:
```

iulianbarbu replied:
Unresolving this. Suggested to also move this to a more appropriate section. LMK what you think.

EleisonC replied:
I have made the appropriate changes @iulianbarbu.  FYI creating the script gh-issue

EleisonC replied:
https://github.com/paritytech/polkadot-sdk/issues/7267

EleisonC replied:
I guess you could assign it to me if you don't mind 😀

iulianbarbu commented on templates/minimal/README.md:
````diff
@@ -37,11 +37,25 @@
+
 - 👤 The template has no consensus configured - it is best for experimenting with a single node network.
 
+For the most accurate chain spec generation,  `staging-chain-spec-builder` or other mentioned binaries and runtimes,
+please try to build them based on the Polkadot-sdk most updated master branch an example:
+  ```bash
+  cargo install --path substrate/bin/utils/chain-spec-builder
+  ```
````
This is a good note, but I wouldn't add it anymore. Latest `staging-chain-spec-builder` is good enough for what this readme needs: https://crates.io/crates/staging-chain-spec-builder

skunert replied:
We also need `--locked` here. Please doublecheck all the install commands. `cargo install` by default does not use the lockfile.

iulianbarbu commented on templates/minimal/README.md:
```diff
@@ -81,15 +95,18 @@ be used to run the minimal template's runtime. `polkadot-omni-node` binary crate
 Please see installation section on [crates.io/omni-node](https://crates.io/crates/polkadot-omni-node).
 
-#### Build `minimal-template-runtime`
+#### Build `minimal-template-runtime` (A reminder to  build on the most updated version of polkadot-sdk master)
```
We can remove the reminder addition. We still waant people to refer to the dedicated minimal-template repository, so building out of `polkadot-sdk` can be done, but isn't the user flow we're looking. This readme will end up in this minimal-template dedicated repo, so we shouldn't refer back to `polkadot-sdk` from there.

iulianbarbu commented on templates/minimal/README.md:
````diff
@@ -81,15 +95,18 @@ be used to run the minimal template's runtime. `polkadot-omni-node` binary crate
 
 #### Install `staging-chain-spec-builder`
+  ```bash
+  cargo install --path substrate/bin/utils/chain-spec-builder
+```
 
-Please see the installation section at [`crates.io/staging-chain-spec-builder`](https://crates.io/crates/staging-chain-spec-builder).
+Or Please see the installation section at [`crates.io/staging-chain-spec-builder`](https://crates.io/crates/staging-chain-spec-builder).
````
Better to revert this because we will assume this readme lives in https://github.com/paritytech/polkadot-sdk-minimal-template.

iulianbarbu commented on templates/parachain/README.md:
```diff
@@ -35,6 +35,11 @@
 
 - ⏫ This template provides a starting point to build a [parachain](https://wiki.polkadot.network/docs/learn-parachains).
 
+- 🔧 To simplify the process of using Zombienet with Omni Node, we’ve added a pre-configured
+development chain spec (dev_chain_spec.json) to the parachain template. Additionally,
+the zombienet-omni-node.toml file has been updated to include a valid chain spec path
+pointing to dev_chain_spec.json
```
Same as for minimal template

iulianbarbu commented on templates/parachain/README.md:
````diff
@@ -43,6 +48,14 @@ such as a [Balances pallet](https://paritytech.github.io/polkadot-sdk/master/pal
 
 - 👉 Learn more about parachains [here](https://wiki.polkadot.network/docs/learn-parachains)
 
+- For the most accurate chain spec generation,  `staging-chain-spec-builder` or other mentioned
+binaries and runtimes,please try to build them based on the Polkadot-sdk most updated master branch 
+an example:
+  ```bash
+  cargo install --path substrate/bin/utils/chain-spec-builder
+  ```
+
+
````
Same as for minimal template

iulianbarbu reviewed (commented):
Hi @EleisonC ! This is very nice and looks about done, except a few things that can be improved.

Other than this I plan to sync the templates with stable2412 release today & hopefully this will work without significant issues. Once that's done you can read every template readme as if it would be part of the dedicated template repos. This is just as FYI, I don't think you need to change more than what I suggested with this review, for this PR.

Other than this, would be great to open an issue related to the script we've discussed about, where we'll apply all this learnings about how to generate reproducible chain specs, if needed.

skunert commented on .github/workflows/misc-sync-templates.yml:
```diff
@@ -27,7 +27,63 @@ on:
+          run: |
+            sudo apt-get update
+            sudo apt-get install -y protobuf-compiler
+            cargo install --git https://github.com/chevdor/srtool-cli
```
Please add `--locked` to these git install commands.

skunert reviewed (commented):
Looking really good, only minor nits!

Thanks again for seeing this through :)!

EleisonC commented:
> Looking really good, only minor nits!
> 
> Thanks again for seeing this through :)!

Thank you for your guidance and feedback! I've addressed the minor nits. 😊

iulianbarbu commented on templates/minimal/README.md:
```diff
@@ -37,11 +37,26 @@
 - 🤏 This template is a minimal (in terms of complexity and the number of components)
 template for building a blockchain node.
```
Resolving this since the same subject is discussed under a more actual comment.

skunert commented on .github/workflows/misc-sync-templates.yml:
```diff
@@ -27,7 +27,66 @@ on:
+        - name: Prepare upload directory
+          run: |
+            mkdir -p artifacts-${{ matrix.template }}
+            cp dev_chain_spec.json artifacts-${{ matrix.package_name }}/dev_chain_spec.json
```
Here you are using `matrix.package_name` instead of `matrix.template`, I think this is a mismatch.

EleisonC replied:
Thanks for the catch. I have addressed it

iulianbarbu commented on .github/workflows/misc-sync-templates.yml:
```diff
@@ -37,6 +96,18 @@ jobs:
       template-path: "polkadot-sdk-${{ matrix.template }}-template"
     steps:
+      # 0. download dev chain spec
+      - uses: actions/checkout@v4
+        with:
+          path: polkadot-sdk
+          ref: "${{ github.event.inputs.stable_release_branch }}"
+
+      - name: Download template artifacts
+        uses: actions/download-artifact@v4
+        with:
+          name: artifacts-${{ matrix.template }}
+          path: templates/${{ matrix.template }}/
+        if: matrix.template != 'solochain'
```
Can we remove the first action checkout here? We're doing it once more below, after setting the git identity. I am thinking we can also move the `Download template artifacts` below the actions checkout that runs after `Configure git identity`. What do you think?

iulianbarbu commented on templates/minimal/README.md:
```diff
@@ -37,11 +37,17 @@
 - 👤 The template has no consensus configured - it is best for experimenting with a single node network.
 
+
+- For the most up-to-date chain specification, refer to [staging-chain-spec-builder](https://crates.io/crates/staging-chain-spec-builder)
```
I would remove this note from the intro. It isn't very relevant from an introduction standpoint for the template, and the latest info for the chain spec isn't necessarily part of `staging-chain-spec-builder`. I would keep the `#Intro` section as before.

You don't need to add a mention about `staging-chain-spec-builder` in other place either. We already have it mentioned in the steps below (e.g. under `Starting a Minimal Template Chain`/`Omni Node`/`Install staging-chain-spec-builder`.

iulianbarbu commented on templates/minimal/README.md:
```diff
@@ -147,6 +153,11 @@ docker run --rm polkadot-sdk-minimal-template
 We can install `zombienet` as described [here](https://paritytech.github.io/zombienet/install.html#installation),
 and `zombienet-omni-node.toml` contains the network specification we want to start.
 
+To simplify the process of starting the minimal template with zombienet and Omni Node, we've added a pre-configured 
+development chain spec (dev_chain_spec.json) to the minimal template. The zombienet-omni-node.toml file of this template
+points to it, but you can update it to an updated path to the new chain spec generated on your machine, as shown below.
```
```suggestion
To simplify the process of starting the minimal template with zombienet and Omni Node, we've added a pre-configured 
development chain spec (dev_chain_spec.json) to the minimal template. The zombienet-omni-node.toml file of this template
points to it, but you can update it to an updated path to the new chain spec generated on your machine.
```

My initial suggestion was to move this right into the section: `Update zombienet-omni-node.toml with a valid chain spec path`. It can be placed right at the beginning of the section, but then I would change the next paragraph: `Before starting...` by removing the first sentence.

EleisonC replied:
I made the changes to this

iulianbarbu commented on templates/parachain/README.md:
````diff
@@ -144,6 +147,12 @@ Download the `polkadot` (and the accompanying `polkadot-prepare-worker` and `pol
 export PATH="$PATH:<path/to/binaries>"
 ```
 
+
+To simplify the process of using the parachain-template with zombienet and Omni Node, we've added a pre-configured
+development chain spec (dev_chain_spec.json) to the parachain template. The zombienet-omni-node.toml file of this
+template points to it, but you can update it to an updated chain spec generated on your machine, as shown below:
````
Exactly the same as for minimal

iulianbarbu commented on .github/workflows/misc-sync-templates.yml:
```diff
@@ -27,7 +27,66 @@ on:
+          - template: parachain
+            package_name: 'parachain-template-runtime'
+            runtime_path: './templates/parachain/runtime'
+            runtime_wasm_path: 'minimal-template-runtime/minimal_template_runtime.compact.compressed.wasm'
```
```suggestion
            runtime_wasm_path: 'parachain-template-runtime/parachain_template_runtime.compact.compressed.wasm'
```

iulianbarbu reviewed (commented):
Just felt we can make the docs a bit clearer, so left a few clarifying comments, besides a few others for the `misc-sync-templates.yml`. It is mostly there :D

EleisonC commented:
Hey @iulianbarbu done with the feedback.

iulianbarbu commented on .github/workflows/misc-sync-templates.yml:
```diff
@@ -27,7 +27,66 @@ on:
+          - template: parachain
+            package_name: 'parachain-template-runtime'
+            runtime_path: './templates/parachain/runtime'
+            runtime_wasm_path: 'parachain-template-runtime/parachain_template_runtime.compact.compressed.wasm'
```
nit:
```suggestion
            runtime_wasm_path: parachain-template-runtime/parachain_template_runtime.compact.compressed.wasm
```

iulianbarbu commented on templates/minimal/README.md:
```diff
@@ -37,11 +37,14 @@
 - 🤏 This template is a minimal (in terms of complexity and the number of components)
 template for building a blockchain node.
 
+
```
nit:
```suggestion
```

iulianbarbu commented on templates/minimal/README.md:
```diff
@@ -37,11 +37,14 @@
 - 🔧 Its runtime is configured with a single custom pallet as a starting point, and a handful of ready-made pallets
 such as a [Balances pallet](https://paritytech.github.io/polkadot-sdk/master/pallet_balances/index.html).
 
+
```
nit:
```suggestion
```

iulianbarbu commented on templates/parachain/README.md:
```diff
@@ -144,8 +145,16 @@ Download the `polkadot` (and the accompanying `polkadot-prepare-worker` and `pol
+
 #### Update `zombienet-omni-node.toml` with a valid chain spec path
 
+To simplify the process of using the parachain-template with zombienet and Omni Node, we've added a pre-configured
+development chain spec (dev_chain_spec.json) to the parachain template. The zombienet-omni-node.toml file of this
+template points to it, but you can update it to an updated chain spec generated on your machine. To generate a
+chain spec refer to [staging-chain-spec-builder](https://crates.io/crates/staging-chain-spec-builder)
+
+Then make the changes in the network specification like so:
+
```
I recalled we added this file for the `parachain-template`: https://github.com/paritytech/polkadot-sdk/blob/master/templates/parachain/README.docify.md, which should help with keeping in sync the README.md with code sections from parachain template components.

To make README.md changes for parachain template now we'll need to update the `README.docify.md` exactly as you updated the README.md, and then run (which will regenerate the README.md):

```bash
cd templates/parachain && cargo check --features generate-readme
```

LMK if you encounter issues with this.

EleisonC replied:
Hey @iulianbarbu I am done with the feedback

iulianbarbu reviewed (approved):
Looks great! Thank you for sticking through this, it wasn't an easy one. :) I'll merge once you make the change for the `README.docify.md`.

iulianbarbu commented:
Some trailing spaces in minimal/README.md make a required check fail. You can fix that and I will try merging after.

@alvicsam just as a headsup, you requested changes, you might want to take a look over this too before merging.
